src/vmm/src/arch/aarch64/regs.rs

```diff
@@ -157,265 +110,733 @@ arm64_sys_reg!(ID_AA64MMFR2_EL1, 3, 0, 0, 7, 2);
-        vm.get_preferred_target(&mut kvi).unwrap();
-
-        // Must fail when vcpu is not initialized yet.
-        let res = get_mpidr(&vcpu);
+}
+
+/// Reference to the aarch64 register.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Aarch64RegisterRefMut<'a> {
+    /// ID of the register
+    pub id: u64,
+    data: &'a mut [u8],
+}
+
+impl<'a> Aarch64RegisterRefMut<'a> {
+    /// Creates new register reference with provided id and data
+    pub fn new(id: u64, data: &'a mut [u8]) -> Self {
         assert_eq!(
-            res.unwrap_err(),
-            Error::GetOneReg(MPIDR_EL1, kvm_ioctls::Error::new(8))
+            reg_size(id) as usize,
+            data.len(),
+            "Attempt to create a register reference with incompatible id and data length"
         );
 
-        vcpu.vcpu_init(&kvi).unwrap();
-        assert_eq!(get_mpidr(&vcpu).unwrap(), 0x8000_0000);
+        Self { id, data }
     }
 
-    #[test]
-    fn test_get_set_regs() {
-        let kvm = Kvm::new().unwrap();
-        let vm = kvm.create_vm().unwrap();
-        let vcpu = vm.create_vcpu(0).unwrap();
-        let mut kvi: kvm_bindings::kvm_vcpu_init = kvm_bindings::kvm_vcpu_init::default();
-        vm.get_preferred_target(&mut kvi).unwrap();
-
-        // Must fail when vcpu is not initialized yet.
-        let mut regs = Vec::new();
-        let res = get_all_registers(&vcpu, &mut regs);
+    /// Returns register size in bytes
+    pub fn size(&self) -> RegSize {
+        reg_size(self.id).into()
+    }
+
+    /// Returns a reference to the register value.
+    /// Type `T` must be of the same length as an
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn value<T>(&self) -> &T {
+        assert_eq!(
+            std::mem::size_of::<T>(),
+            self.data.len(),
+            "Attempt to cast aarch64 register to incorrect type"
+        );
+
+        // # Safety:
+        // We checked that size of `T` is same as `data.len`
+        unsafe { &*(self.data.as_ptr().cast::<T>()) }
+    }
+
+    /// Returns a mutable reference to the register value.
+    /// Type `T` must be of the same length as an
+    /// underlying data slice. Otherwise this method
+    /// will panic.
+    pub fn value_mut<T>(&mut self) -> &mut T {
+        assert_eq!(
+            std::mem::size_of::<T>(),
+            self.data.len(),
+            "Attempt to cast aarch64 register to incorrect type"
+        );
+
+        // # Safety:
+        // We checked that size of `T` is same as `data.len`
+        unsafe { &mut *(self.data.as_mut_ptr().cast::<T>()) }
+    }
+
+    /// Returns register data as a byte slice
+    pub fn as_slice(&self) -> &[u8] {
+        self.data
+    }
+}
+
+/// Struct describing a aarch64 register.
+#[derive(Debug, Clone, Copy, Versionize)]
+pub struct Aarch64RegisterOld {
+    /// ID of the register.
+    pub id: u64,
+    /// Register data.
+    pub data: u128,
+}
+
+impl Aarch64RegisterOld {
```

Comment:
Would `Fixed` be a better name than `Old`? Old is a relative term.
Also, can we add a deprecation note that it needs not be used in new code?